tests/modules/images/image-loading-optimization/optimization-tests.php

```diff
@@ -0,0 +1,528 @@
+<?php
+/**
+ * Tests for image-loading-optimization module optimization.php.
+ *
+ * @package performance-lab
+ * @group   image-loading-optimization
+ */
+
+class ILO_Optimization_Tests extends WP_UnitTestCase {
+
+	public function tear_down() {
+		parent::tear_down();
+		unset( $_SERVER['REQUEST_URI'] );
+	}
+
+	/**
+	 * Test ilo_maybe_add_template_output_buffer_filter().
+	 *
+	 * @covers ::ilo_maybe_add_template_output_buffer_filter
+	 */
+	public function test_ilo_maybe_add_template_output_buffer_filter() {
+		$this->assertFalse( has_filter( 'ilo_template_output_buffer', 'ilo_optimize_template_output_buffer' ) );
+
+		add_filter( 'ilo_can_optimize_response', '__return_false', 1 );
+		ilo_maybe_add_template_output_buffer_filter();
+		$this->assertFalse( ilo_can_optimize_response() );
+		$this->assertFalse( has_filter( 'ilo_template_output_buffer', 'ilo_optimize_template_output_buffer' ) );
+
+		add_filter( 'ilo_can_optimize_response', '__return_true', 2 );
+		$this->go_to( home_url( '/' ) );
+		$this->assertTrue( ilo_can_optimize_response() );
+		ilo_maybe_add_template_output_buffer_filter();
+		$this->assertSame( 10, has_filter( 'ilo_template_output_buffer', 'ilo_optimize_template_output_buffer' ) );
+	}
+
+	/**
+	 * Data provider.
+	 *
+	 * @return array
+	 */
+	public function data_provider_test_ilo_can_optimize_response(): array {
+		return array(
+			'homepage'           => array(
+				'set_up'   => function () {
+					$this->go_to( home_url( '/' ) );
+				},
+				'expected' => true,
+			),
+			'homepage_filtered'  => array(
+				'set_up'   => function () {
+					$this->go_to( home_url( '/' ) );
+					add_filter( 'ilo_can_optimize_response', '__return_false' );
+				},
+				'expected' => false,
+			),
+			'search'             => array(
+				'set_up'   => function () {
+					self::factory()->post->create( array( 'post_title' => 'Hello' ) );
+					$this->go_to( home_url( '?s=Hello' ) );
+				},
+				'expected' => false,
+			),
+			'customizer_preview' => array(
+				'set_up'   => function () {
+					$this->go_to( home_url( '/' ) );
+					global $wp_customize;
+					/** @noinspection PhpIncludeInspection */
+					require_once ABSPATH . 'wp-includes/class-wp-customize-manager.php';
+					$wp_customize = new WP_Customize_Manager();
+					$wp_customize->start_previewing_theme();
+				},
+				'expected' => false,
+			),
+			'post_request'       => array(
+				'set_up'   => function () {
+					$this->go_to( home_url( '/' ) );
+					$_SERVER['REQUEST_METHOD'] = 'POST';
+				},
+				'expected' => false,
+			),
+		);
+	}
+
+	/**
+	 * Test ilo_can_optimize_response().
+	 *
+	 * @covers ::ilo_can_optimize_response
+	 * @dataProvider data_provider_test_ilo_can_optimize_response
```

Comment:
```suggestion
	 * @covers ::ilo_can_optimize_response
	 *
	 * @dataProvider data_provider_test_ilo_can_optimize_response
```
Nice if we add @param that used in dataProvider.